slapo/op/fused_bias.py

```diff
@@ -118,3 +109,52 @@ def reset_parameters(self, prev_weight=None):
 
     def forward(self, inp):
         return self.func(inp, self.bias)
+
+
+def bias_dropout(
+    x: torch.Tensor,
+    bias: torch.Tensor,
+    p: float = 0.5,
+    training: bool = True,
+    inplace: bool = False,
+) -> torch.Tensor:
+    return F.dropout(x + bias, p=p, training=training, inplace=inplace)
+
+
+class FusedBiasDropout(torch.nn.Module):
+    def __init__(
+        self,
+        size,
+        p=0.5,
+        training=True,
+        inplace=False,
+        device=None,
+        dtype=None,
+        prev_weight=None,
+    ):
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__()
+        self.bias = torch.nn.Parameter(torch.empty(size, **factory_kwargs))
+        self.reset_parameters(prev_weight)
+        self.p = p
+        self.training = training
+        self.inplace = inplace
+        self.func = torch.jit.script(bias_dropout)
```

Comment:
No. The torchscript function can be used for training too. Otherwise the `.fuse` primitive won't work neither.

Reply:
That is the trickiest part. If you only pass in a function to TorchScript, it will not optimize the backward pass. To let TorchScript capture the backward pass, the computation needs to be encapsulated in a `nn.Module`.

Reply:
For `.fuse()` primitive, I always pass in a `GraphModule`, which is also a `nn.Module`, to TorchScript, so it is able to fuse the operations in the backward pass.

Reply:
I see. That's tricky. But the current bias_gelu is implemented in the same way. Does that mean it only affect the backward performance?

Reply:
Nope. We have the [backward function](https://github.com/awslabs/slapo/blob/cd22329a4344187347e08a3355c89a01b9ca636f/slapo/op/bias_gelu.py#L31-L39) for `bias_gelu`. This should work. I remember there will be ~3% performance degradation if we only fuse the forward op of `bias_gelu`. I'm not sure how much will be affected for `bias_dropout`.

Reply:
Sorry I meant `bias_new_gelu`...so I guess this is another reason of using `memory_efficient_fusion`, because it will perform AOTAugograd and also optimize the backward graph.

Reply:
Make sense. But for now, we can encapsulate the function in a module as a workaround.

Reply:
And yeah, I checked the implementation of `bias_new_gelu`. It indeed can lead to possible performance degradation. Do we have any models using `FusedBiasNewGELU`?

Reply:
The default config of GPT2 and GPT-Neo use it.

Reply:
Oh I guess it might be ok because I used `memory_efficient_fusion` in BiasNewGeLU...

Reply:
Emm. The 2% performance gap between Nemo-Megatron...

Reply:
Did you use pytorch 2.0 when benchmarking the GPT2 model?

Reply:
This API is available in 1.13 already.